### PR TITLE
[WIP] add resource to ownerreference

### PR DESCRIPTION
**Pull request by deads2k**
Implements https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190607-owner-ref.md .

Collecting a working set here, am willing to split for review.

/kind feature
@kubernetes/sig-api-machinery-api-reviews 
/priority important-soon

```release-note
Adds `resource` to ownerreferences.  See https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190607-owner-ref.md for details.
```

### CODE REVIEW

File: pkg/controller/garbagecollector/garbagecollector.go [hunk 1]
```diff
@@ -339,6 +340,26 @@ func (gc *GarbageCollector) isDangling(reference metav1.OwnerReference, item *no
 		return false, nil, err
 	}
 
+	if reference.Namespace != nil {
```
Comment (caesarxuchao) [inline pkg/controller/garbagecollector/garbagecollector.go]: I thought we were going to make the gc skip gc.apiResource if the reference.Namespace and reference.Resource are specified, but the implementation here is using reference.Namespace and reference.Resource for additional validation.

**Comment by caesarxuchao** [comment]
I will be OOO for a month. I'll leave it to @cheftako for additional review.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 2]
```diff
@@ -341,10 +341,14 @@ type OwnerReference struct {
...
 	BlockOwnerDeletion *bool `json:"blockOwnerDeletion,omitempty" protobuf:"varint,7,opt,name=blockOwnerDeletion"`
-	// Resource of the referent.  If specified, it takes priority over `kind`.
+	// resource of the referent.  If specified, it takes priority over `kind`.
 	// More info: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#resources
 	// +optional
 	Resource string `json:"resource,omitempty" protobuf:"bytes,8,opt,name=resource"`
+	// namespace of the referent.  If specified, it must be empty for cluster scoped resources or in the same namespace as the dependent.
+	// More info: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	Namespace *string `json:"namespace,omitempty" protobuf:"bytes,9,opt,name=namespace"`
```
[the remark this answers is not in the document]
Comment (deads2k) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]: > Why do we need to add the Namespace field? Is the purpose avoiding checking with the server if the referent is cluster scoped or namespace scoped?
> 
> We should update the KEP if we want to add this field to the API.

Yes, that's exactly why.  I will open an update to the KEP, but I wanted to prove the end to end flow before updating with the new cases.

File: staging/src/k8s.io/apimachinery/pkg/api/validation/objectmeta_test.go [hunk 3]
```diff
@@ -124,6 +138,20 @@ func TestValidateObjectMetaOwnerReferences(t *testing.T) {
...
+					APIVersion: "v1",
+					Kind:       "Pod",
+					Resource:   "tricky/resource",
+					Name:       "name",
+					UID:        "1",
+				},
+			},
+			expectError:          true,
+			expectedErrorMessage: "a DNS-1035 label must consist of lower case alphanumeric",
+		},
```
[the remark this answers is not in the document]
Comment (deads2k) [reply staging/src/k8s.io/apimachinery/pkg/api/validation/objectmeta_test.go]: > Can you add a case for namespace mismatching?

Yes, certainly will.

File: pkg/controller/garbagecollector/garbagecollector.go (hunk 1 above)
Comment (deads2k) [reply pkg/controller/garbagecollector/garbagecollector.go]: > I thought we were going to make the gc skip gc.apiResource if the reference.Namespace and reference.Resource are specified, but the implementation here is using reference.Namespace and reference.Resource for additional validation.

Yes.  I will update, but perhaps in a follow-on pull.  I'll probably also leave these checks in place for cases where one or the other, but not both are specified and we want the cross check.

**Pull request closed without merging**